File: lc/BinarySearch/B15ArrTwoSum.java
package lc.BinarySearch;
// https://leetcode.com/explore/learn/card/binary-search/144/more-practices/1035/
class SolutionC05ArrTwoSum {
    public int[] twoSum(int[] nums, int target) {
        int low = 0, high = nums.length - 1;
        while(high > low) {
            System.out.println("low: "+low+" high: "+high);
            int sum = nums[low] + nums[high];
            if(sum == target) {
                int res[] = {low+1, high+1};
                System.out.println("---------");
                return res;
            }else if( sum > target ) {
                high--;
            }else {
                low++;
            }
        }
        int res[] = {-1,-1};
        return res;
    }
}

class B15ArrTwoSum {
    private static boolean isSame(int[] ans, int[] expected){
        if(ans.length != expected.length) return false;
        for(int i=0;i<ans.length; i++){
            if(ans[i] != expected[i]) return false;
        }
        return true;
    }
    public static void main(String[] args){
        SolutionC05ArrTwoSum sol = new SolutionC05ArrTwoSum();

        int tc1[] = {2,7,11,15};
        int target1 = 9;
        int res1[] = {1,2};

        int tc2[] = {2,3,4};
        int target2 = 6;
        int res2[] = {1,3};

        int tc3[] = {-1,0};
        int target3 = -1;
        int res3[] = {1,2};

        assert isSame(sol.twoSum(tc1, target1), res1);
        assert isSame(sol.twoSum(tc2, target2), res2);
        assert isSame(sol.twoSum(tc3, target3), res3);
    }
}

File: lc/BinarySearch/B15ArrTwoSum.js
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(numbers, target) {
    function search(target2, start, end) {
        if(start > end || start >= numbers.length || end < 0) {
            return -1;
        }
        const mid = Math.floor(start + (end-start)/2);
        if(numbers[mid] === target2) {
            return mid;
        } else if(numbers[mid] > target2) {
            return search(target2, start, mid - 1);
        }
        return search(target2, mid + 1, end);
    }
    for(let i=0;i<numbers.length;i++) {
        let foundIndex = -1;
        const remaining = target - numbers[i];
        if(remaining <= numbers[i]) {
            foundIndex = search(remaining, 0, i-1);
        }
        if(remaining >= numbers[i] && foundIndex === -1) {
            foundIndex = search(remaining, i+1, numbers.length-1);
        }
        if(foundIndex !== -1) {
            return [i+1 , foundIndex + 1].sort();
        }
    }
};

console.assert(twoSum([2,3,4], 6));
console.assert(twoSum([2,7,11,15], 6));
console.assert(twoSum([-1,0], 6));
